examples/twittermap/web/public/javascripts/timeseriescache/services.js

```diff
@@ -0,0 +1,128 @@
+/*
+ * This module caches time-series histogram data. The common services module communicates with
+ * middle ware and the time-series cache.
+ */
+'use strict';
+angular.module('cloudberry.timeseriescache', [])
+    .service('TimeSeriesCache', function () {
+
+        // The key-value stores time series results of a query.
+        var timeseriesStore = new HashMap();
+        // To check if keyword in query is changed (new query?)
+        var currentKeywords = [""];
+        // To check if geo level in query is changed.
+        var currentGeoLevel = "";
+        // To check if time range in query changed (new query?)
+        var endDate = new Date();
+        // Deducts -1 to create a default value that can't exist
+        endDate.setDate(endDate.getDate() - 1);
+        // Cached time interval of complete byTimeRequest result.
+        var cachedTimeRange = {
+            start: new Date(),
+            end: endDate
+        };
+        const INVALID_VALUE = 0;
+
+        /**
+         * Checks keyword, time range and the cache store, and returns the geoIds that
+         * are not already cached.
+         */
+        this.getGeoIdsNotInCache = function (keywords, timeInterval, geoIds, geoLevel) {
+            // The length of geoIdsNotInCache is 0 in case of complete cache hit,
+            // same length as the geoIds parameter in case of complete cache miss,
+            // otherwise in range (0, geoIds.length)
+            var geoIdsNotInCache = [];
+            // New query case
+            if (keywords.toString() != currentKeywords.toString() ||
+                timeInterval.start < cachedTimeRange.start ||
+                timeInterval.end > cachedTimeRange.end ||
+                geoLevel != currentGeoLevel) {
+                timeseriesStore.clear();
+                currentKeywords = keywords.slice();
+                currentGeoLevel = geoLevel;
+                cachedTimeRange.start = timeInterval.start;
```

Comment:
Let's make sure the assignment is not shallow copy, but should be deep copy. Otherwise, the `cachedTimeRange.start` will always equal to `timeInterval.start`, similar to `end` and `currentKeywords`.